sqlmesh/core/plan/builder.py

```diff
@@ -650,6 +709,85 @@ def _ensure_new_env_with_changes(self) -> None:
                 "No changes were detected. Make a change or run with --include-unmodified to create a new environment without changes."
             )
 
+    def _upstream_directly_modified(
+        self, s_id: SnapshotId, directly_modified: t.Set[SnapshotId], dag: DAG[SnapshotId]
+    ) -> t.List[SnapshotId]:
+        return [
+            upstream_id for upstream_id in dag.upstream(s_id) if upstream_id in directly_modified
+        ]
+
+    def _validate_schema_change(
+        self,
+        s_id: SnapshotId,
+        upstream_snapshot_ids: t.List[SnapshotId] = [],
+        force_warn: bool = False,
+    ) -> None:
+        """Determine if a schema change requires a DROP operation"""
+        s_id_snapshot = t.cast(Snapshot, self._context_diff.snapshots.get(s_id))
+
+        if not s_id_snapshot.model.on_schema_change.is_ignore or force_warn:
+            schema_differ_args = getattr(
+                DIALECT_TO_ENGINE_ADAPTER[s_id_snapshot.model.dialect], "schema_differ_args"
+            )
+            schema_differ = SchemaDiffer(**schema_differ_args)
```

Comment:
why do we need to instantiate this here? isn't this already an attribute of the engine adapter?